datalad/support/gitrepo.py

```diff
@@ -2836,7 +2836,7 @@ def set_gitattributes(self, attrs, attrfile='.gitattributes', mode='a'):
                         attrline += ' -{}'.format(a)
                     else:
                         attrline += ' {}={}'.format(a, val)
-                f.write('\n{}'.format(attrline))
+                f.write('{}\n'.format(attrline))
```

Comment:
Unfortunately it wouldn't be as safe: default mode is 'a' so we aren't rewriting the file but just adding a new record. What is it lacked final new line to start with?

Reply:
But may be we could avoid writing an initial new line if eg file didn't exist or had size 0

Reply:
We could check whether the file (if it exists) ends with \n. If not, start with \n.

Reply:
could be a windows one AFAIK (was it `\r\n` or `\n\r`? has been a while) so care should be taken.
worth `git blame` and looking at the original comments/PRs on motivation why we have done it this way

Reply:
an alternative - `if mode=='a'` `readlines` it in full first (and start with empty list otherwise), prepare all extra lines, and then write it out consistently in full.  I do not think that performance hit would be notable... again -- worth checking original PR(s) on either such idea was ruled out for some reason (beyond "performance" which is IMHO is nonissue)

Reply:
This was the motivation behind the change. https://github.com/datalad/datalad/pull/2912#issuecomment-429224540

Windows is `\r\n`, so there shouldn't be an issue with my suggestion.

Reply:
I'm not familiar with how datalad tests, but it seems there is no test for the case where file exists and last character isn't `\n`?